crates/indexer/src/runner.rs

```diff
@@ -0,0 +1,103 @@
+use gw_web3_rpc_client::{convertion::to_l2_block, godwoken_rpc_client::GodwokenRpcClient};
+use rust_decimal::{prelude::ToPrimitive, Decimal};
+use sqlx::PgPool;
+
+use crate::{config::IndexerConfig, Web3Indexer};
+use anyhow::{anyhow, Result};
+
+pub struct Runner {
+    indexer: Web3Indexer,
+    pg_pool: PgPool,
+    local_tip: Option<u64>,
+    godwoken_rpc_client: GodwokenRpcClient,
+}
+
+impl Runner {
+    pub fn new(config: IndexerConfig, pg_pool: PgPool) -> Result<Runner> {
+        let indexer = Web3Indexer::new(
+            pg_pool.clone(),
+            config.l2_sudt_type_script_hash,
+            config.polyjuice_type_script_hash,
+            config.rollup_type_hash,
+            config.eth_account_lock_hash,
+            config.tron_account_lock_hash,
+            config.godwoken_rpc_url.as_str(),
+        );
+        let godwoken_rpc_client = GodwokenRpcClient::new(config.godwoken_rpc_url.as_str());
+        let runner = Runner {
+            indexer,
+            local_tip: None,
+            pg_pool,
+            godwoken_rpc_client,
+        };
+        Ok(runner)
+    }
+
+    // None means no local blocks
+    pub async fn tip(&self) -> Result<Option<u64>> {
+        let tip = match self.local_tip {
+            Some(t) => Some(t),
+            None => self.get_db_tip_number().await?,
+        };
+        Ok(tip)
+    }
+
+    pub async fn bump_tip(&mut self) -> Result<()> {
+        match self.local_tip {
+            None => {
+                self.local_tip = if let Some(n) = self.get_db_tip_number().await? {
+                    Some(n)
+                } else {
+                    Some(0)
+                }
+            }
+            Some(t) => {
+                self.local_tip = Some(t + 1);
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn get_db_tip_number(&self) -> Result<Option<u64>> {
+        let row: Option<(Decimal,)> =
+            sqlx::query_as("select number from blocks order by number desc limit 1;")
+                .fetch_optional(&self.pg_pool)
+                .await?;
+        let num = row.and_then(|(n,)| n.to_u64());
+        Ok(num)
+    }
+
+    pub async fn insert(&mut self) -> Result<bool> {
+        let local_tip = self.tip().await?;
```

Comment:
This fn doesn't seem to handle the layer 2 block forking case

Reply:
Yes, will draft a new pr to do this.